sdl-snowflake/src/main/scala/io/smartdatalake/workflow/dataframe/snowflake/SnowparkSubFeed.scala

```diff
@@ -0,0 +1,225 @@
+  }
+  override def applyExecutionModeResultForOutput(result: ExecutionModeResult)(implicit context: ActionPipelineContext): SnowparkSubFeed = {
+    this.copy(partitionValues = result.inputPartitionValues, filter = result.filter, isSkipped = false, dataFrame = None)
+  }
+  override def withDataFrame(dataFrame: Option[GenericDataFrame]): SnowparkSubFeed = this.copy(dataFrame = dataFrame.map(_.asInstanceOf[SnowparkDataFrame]))
+  override def withPartitionValues(partitionValues: Seq[PartitionValues]): DataFrameSubFeed = this.copy(partitionValues = partitionValues)
+  override def asDummy(): SnowparkSubFeed = this.copy(isDummy = true)
+  override def withFilter(partitionValues: Seq[PartitionValues], filter: Option[String]): DataFrameSubFeed = {
+    this.copy(partitionValues = partitionValues, filter = filter)
+      .applyFilter
+  }
+
+  override def isStreaming: Option[Boolean] = Some(false) // no spark streaming with Snowpark
+
+  override def movePartitionColumnsLast(partitions: Seq[String]): SnowparkSubFeed = {
+    withDataFrame(dataFrame.map(x => x.movePartitionColsLast(partitions)))
+  }
+}
+
+object SnowparkSubFeed extends DataFrameSubFeedCompanion {
+  override def fromSubFeed(subFeed: SubFeed)(implicit context: ActionPipelineContext): SnowparkSubFeed = {
+    subFeed match {
+      case snowparkSubFeed: SnowparkSubFeed => snowparkSubFeed
+      case dataFrameSubFeed: DataFrameSubFeed =>
+        val convertedSchema = dataFrameSubFeed.schema.map(_.convert(subFeedType))
+        SnowparkSubFeed(convertedSchema.map(s => getEmptyDataFrame(s, subFeed.dataObjectId)), subFeed.dataObjectId, subFeed.partitionValues, subFeed.isDAGStart, subFeed.isSkipped, isDummy = convertedSchema.isDefined)
+      case _ => SnowparkSubFeed(None, subFeed.dataObjectId, subFeed.partitionValues, subFeed.isDAGStart, subFeed.isSkipped)
+    }
+  }
+  override def subFeedType: universe.Type = typeOf[SnowparkSubFeed]
+  override def col(colName: String): SnowparkColumn = {
+    SnowparkColumn(functions.col(colName))
+  }
+  override def lit(value: Any): SnowparkColumn = {
+    SnowparkColumn(functions.lit(value))
+  }
+  override def max(column: GenericColumn): SnowparkColumn = {
+    column match {
+      case snowparkColumn: SnowparkColumn => SnowparkColumn(functions.max(snowparkColumn.inner))
+      case _ => throw new IllegalStateException(s"Unsupported subFeedType ${column.subFeedType.typeSymbol.name} in method max")
+    }
+  }
+  override def count(column: GenericColumn): SnowparkColumn = {
+    column match {
+      case snowparkColumn: SnowparkColumn => SnowparkColumn(functions.count(snowparkColumn.inner))
+      case _ => throw new IllegalStateException(s"Unsupported subFeedType ${column.subFeedType.typeSymbol.name} in method count")
+    }
+  }
+  override def explode(column: GenericColumn): SnowparkColumn = {
+    // TODO: check if this can be done by a udf?
+    throw new NotImplementedError("explode array is not implemented in Snowpark")
```

Comment:
I found this implementation in snowflake DataFrame API:
com.snowflake.snowpark.DataFrame#flatten
https://docs.snowflake.com/en/sql-reference/functions/flatten.html
I think it does the same thing, but it needs a DataFrame...

Reply:
You're right, the function exists on DataFrame-level and not on Column-level. In my opinion this is due the fact, that Snowflake treats nested datatypes as variant in SQL. And as Snowpark gets translated to SQL, it's difficult to implement a flatten function on column level, which would need to parse the variant and apply flatten at some level, as the schema is not known at plan-time, but only at execution-time...
I think Spark is more sophisticated with nested data types than Snowpark. Spark knows the schema at plan-time, but in a Snowpark UDF its schema-on-read, means you get a variant and try to cast it to what ever you want:
https://docs.snowflake.com/en/developer-guide/snowpark/reference/scala/com/snowflake/snowpark/types/Variant.html
This also means there are no nested Structs, but only nested Maps...